### ISSUE TITLE
colwise function

### ISSUE THREAD

**Issue by musm** [JuliaStats/Distances.jl#138]
Why is there only a  `colwise` function and not also a  matching rowwise function?
Personally I find `pairwise` better to use, so I'd also support deprecating colwise.

**Comment by dkarrasch** [comment]
But `colwise` and `pairwise` have very different meanings here, independent of whether data points are viewed as columns or rows. `colwise` is meant to compute distances only between corresponding data points/columns. `pairwise` is meant to compute all pairwise distances, and can be used to compute the distances between matrices with different number of data points (say, columns), which is impossible for `colwise`, except for when one of the two data sets is a single data point, i.e., a vector.

**Comment by musm** [comment]
Good point regarding pairwise (comment edited). I think I was getting at a `rowwise` that computes the distances between rows of the matrixes i.e. `rowwise(dist, X,Y) = colwise(dist, X', Y')

**Comment by johnnychen94** [comment]
I suppose one of reason here is because julia matrix is stored in column-major order, so a row-wise loop might raise performance issue.

```julia
julia> x = [1 2 3; 4 5 6; 7 8 9]
3×3 Array{Int64,2}:
 1  2  3
 4  5  6
 7  8  9

julia> x[:]
9-element Array{Int64,1}:
 1
 4
 7
 2
 5
 8
 3
 6
 9
```

**Comment by nalimilan** [comment]
With https://github.com/JuliaLang/julia/pull/32310 we should probably drop `colwise(dist, x, y)` in favor of `map(d, eachol(x), eachcol(y))`. Cc: @simonbyrne

**Comment by dkarrasch** [comment]
I guess we should keep `colwise(dist, x, y)`, make `map(d, eachcol(x), eachcol(y))` the default, but allow specialized methods to optimize for performance.
```julia
using Distances, BenchmarkTools
d = Euclidean(); a = rand(5, 100); b = rand(5, 100);
@btime map($d, $(eachcol(a)), $(eachcol(b))); # 3.090 μs (304 allocations: 13.45 KiB)
@btime colwise($d, $a, $b); # 380.907 ns (1 allocation: 896 bytes)
```

**Comment by simonbyrne** [comment]
@dkarrasch the point of https://github.com/JuliaLang/julia/pull/32310 is that you could write specialized versions that can leverage the memory layout: in this case, you could do:
```julia
Base.map(d::Distance, a::EachCol, b::EachCol) = colwise(d, parent(a), parent(b))
```

**Comment by mroavi** [comment]
I have a question related to the original issue. Suppose I'm given a matrix in which each row corresponds to a point in a 2-dimensional plane, e.g. `rand(100,2)`. I also have another smaller matrix containing 3 points, `rand(3,2)`. For each of the 3 points in the smaller matrix, I need to calculate the distance with each point in the bigger matrix. How can I compute this efficiently?

I noticed that if, on the other hand, I would have been given the transposed version of these matrices, then I could easily calculate this with:
```julia
julia> a = rand(2,100);
julia> b = rand(2,3);
julia> pairwise(SqEuclidean(), a, b, dims=2)
```

Is there a counterpart that instead of working with columns, works with rows?
